File: src/main/java/me/sargunvohra/svlib/capability/SimpleCapabilityProvider.java
package me.sargunvohra.svlib.capability;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import lombok.Getter;
import lombok.val;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import net.minecraftforge.common.util.LazyOptional;

/**
 * Simple provider for use with capabilities that simply attach without complex logic.
 *
 * @param <I> The capability interface
 */
@Getter
public class SimpleCapabilityProvider<I> implements ICapabilityProvider {

  private final Capability<I> type;
  private final I instance;

  public SimpleCapabilityProvider(Capability<I> type) {
    this.type = type;
    this.instance = type.getDefaultInstance();
  }

  @Nonnull
  @Override
  public <T> LazyOptional<T> getCapability(@Nonnull Capability<T> type, @Nullable EnumFacing side) {
    if (type != this.type) return LazyOptional.empty();

    //noinspection NullableProblems
    val ret = LazyOptional.of(this::getInstance);

    // noinspection unchecked
    return (LazyOptional<T>) ret;
  }
}

File: src/main/java/me/sargunvohra/svlib/capability/SimpleCapabilityProvider.kt
package me.sargunvohra.svlib.capability

import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.ICapabilityProvider
import net.minecraftforge.common.util.LazyOptional

/**
 * Simple provider for use with capabilities that simply attach without complex logic or
 * serialization.
 */
open class SimpleCapabilityProvider<Handler>(
    val capability: Capability<Handler>
) : ICapabilityProvider {

    val instance: Handler = capability.defaultInstance!!

    override fun <T> getCapability(capability: Capability<T>, side: EnumFacing?): LazyOptional<T> {
        if (capability !== this.capability)
            return LazyOptional.empty()

        @Suppress("UNCHECKED_CAST")
        return LazyOptional.of { this.instance } as LazyOptional<T>
    }
}
